sdks/java/io/parquet/src/main/java/org/apache/beam/sdk/io/parquet/ParquetIO.java

```diff
@@ -746,6 +780,31 @@ public void processElement(
             tracker.currentRestriction().getFrom(),
             tracker.currentRestriction().getTo());
         Configuration conf = getConfWithModelClass();
+
+        if (useProtoReader) {
+          // Use ProtoParquetReader to read protobuf data.
+          // Derive a Hadoop Path from the file metadata. Adjust as needed.
+          Path path = new Path(file.getMetadata().resourceId().toString());
+
+          // Get the configuration and set the property using the literal.
+          Configuration conf2 = getConfWithModelClass();
+          conf2.set("parquet.proto.ignore.unknown.fields", "TRUE");
+
+          // Use the builder overload that takes a ReadSupport and a Path.
+          try (ParquetReader<GenericRecord> reader =
+              ProtoParquetReader.<GenericRecord>builder(new ProtoReadSupport(), path).build()) {
+            GenericRecord message;
+            while ((message = reader.read()) != null) {
+              // Cast through Object so that parseFn (which expects GenericRecord)
+              // can accept the DynamicMessage.
+              outputReceiver.output(parseFn.apply((GenericRecord) (Object) message));
+            }
+          }
+
+          return; // exit after using the proto path
```

Comment:
For readability, I would recommend structure the code, e.g. 

getParquetFileReader() {
  switch (FORMAT) {
    case AVRO: return getParquetAvroReader()
    case PROTO: return getParquetProtoReader()
  }
}
 instead of branching in place